703. Kth Largest Element in a Stream/KthLargestElementInAStreamStep2.java

```diff
@@ -0,0 +1,192 @@
+    private int[] heap;
+    private int lastHasValueIndex;
+    
+//    public IntMinHeap(int size) {
+//        this.heap = new int[size];
+//        lastHasValueIndex = -1;
+//    }
+    
+    public IntMinHeap(int[] nums) {
+        lastHasValueIndex = nums.length - 1;
+        heapify(nums);
+    }
+    
+    private void heapify(int[] nums) {
+        heap = nums;
+        /*
+         * 子がいない最後の親ノードから、葉ノードと比較しながら根ノードの最小値を更新して、これを根まで続ければ、
+         * 最小のノードから順に根に来る。子がいない最後の親ノードは、計算すると 要素の数n // 2 - 1 インデックス目。 
+         * 親ノードは最低1つの子ノードを持ち(2 * i + 1)、2 * i + 1 < n を満たすと、2つ子ノードを持っている場合も同じ親ノードを指せる
+         * 2 * i + 2 < n で計算しても同じ結果になる
+         */
+        for (int index = nums.length / 2 - 1; index >= 0; index--) {
+            shiftToLeaf(index);
+        }
+    }
+    
+    public void push(int value) {
+        lastHasValueIndex++;
+        heap[lastHasValueIndex] = value;
+        shiftToRoot(lastHasValueIndex);
+    }
+    
+    public int pop() {
+        int poppedValue = heap[0];
+        heap[0] = heap[lastHasValueIndex];
+        heap[lastHasValueIndex] = 0;
+        lastHasValueIndex--;
+        shiftToLeaf(0);
+        return poppedValue;
+    }
+    
+    private void shiftToLeaf(int index) {
+        int rootValue = heap[index];
+        while (index < lastHasValueIndex) {
+            int minChildIndex = getMinChildIndex(index);
+            if (minChildIndex < 0) {
+                break;
+            }            
+            if (heap[minChildIndex] >= rootValue) {
+                break;
+            }
+            heap[index] = heap[minChildIndex];
+            index = minChildIndex;           
+        }
+        heap[index] = rootValue;
+    }
+    
+    private int getMinChildIndex(int index) {
+        int leftChildIndex = 2 * index + 1;
+        if (leftChildIndex > lastHasValueIndex) {
+            return -1;
+        }
+        if (leftChildIndex == lastHasValueIndex) {
+            return leftChildIndex;
+        }
+        int rightChildIndex = 2 * index + 2;
+        if (heap[rightChildIndex] < heap[leftChildIndex]) {
+            return rightChildIndex;
+        }
+        return leftChildIndex;
+    }
+    
+    private void shiftToRoot(int index) {
+        int newValue = heap[index];
+        int parentIndex = getParentIndex(index);
+        while (index > 0 && heap[parentIndex] > newValue) {
+            heap[index] = heap[parentIndex]; 
+            index = parentIndex;      
+            parentIndex = getParentIndex(index);
+        }
+        heap[index] = newValue;
+    }
+
+    private int getParentIndex(int index) {
+        return (index - 1) / 2;
+    }
+    
+    public int getSize() {
+        return lastHasValueIndex + 1;
+    }
+    
+    public int peek() {
+        return heap[0];
+    }
+    
+    public boolean isFull() {
+        return lastHasValueIndex == heap.length - 1;
+    }
+    
+    public void grow() {
+        int[] updatedHeap = Arrays.copyOf(heap, heap.length + 1);

```

Comment:
https://en.wikipedia.org/wiki/Dynamic_array
> To avoid incurring the cost of resizing many times, dynamic arrays resize by a large amount, such as doubling in size, and use the reserved space for future expansion. The operation of adding an element to the end might work as follows:

Reply:
あ、見落としてました。1つずつ増やしていくと1回fullになったらその後毎回の追加のタイミングでcopyが発生するのでpushの処理に余分にコピーの分だけO(n)が掛かり、トータルで自作のpush関数のTime ComplexityがO(nlogn)になりますね。

2倍ずつなどの戦略を取るとlogn回ごとに計算量O(n)のコピー処理が走るので、ならすとO(1)で要素を追加することができます。

Reply:
いや、計算量はO(n)か

Reply:
> 2倍ずつなどの戦略を取るとlogn回ごとに計算量O(n)のコピー処理が走るので、ならすとO(1)で要素を追加することができます。

O(n)の処理をlogn回すると、トータルでO(nlogn)になりませんか？

Reply:
https://en.wikipedia.org/wiki/Geometric_series
![image](https://github.com/goto-untrapped/Arai60/assets/167995483/d009826e-6a7d-43ab-8721-306891a5bafa)

Reply:
ありがとうございます。結論的にはO(1)なので、途中のロジックがおかしかったですね。
雑に処理の回数をn*lognとするのでなくて1 + 2 + 4 + ... 2^(logn)と足していくのが正しいですね。

a=1, r = 2, n = log2(n) + 1なので2nですね。

Reply:
ありがとうございます。
コピー回数を減らすために、コピー時はサイズを2倍にするといった手法が使われるのですね。
また、サイズを1ずつ拡張すると、O(n)になりますね・・・。
コピーごとに2倍のサイズにすることで、O(n)の処理をlogn回すればよくなり、1 + 2 + 4 + ... 2^(logn) = O(2n) = O(n)になると理解しました。

> ならすとO(1)で要素を追加することができます。

上記が分からなかったです・・・。下記のpush関数で要素を追加する場合、O(1)になるということでしょうか？
```
public void push(int value) {
    if (isFull()) {
        grow();
    }
    lastHasValueIndex++;
    heap[lastHasValueIndex] = value;
    shiftToRoot(lastHasValueIndex);
}
public void grow() {
    int[] updatedHeap = Arrays.copyOf(heap, heap.length * 2);
    heap = updatedHeap;
}
```

Reply:
amortized time complexityのことです。
https://en.wikipedia.org/wiki/Amortized_analysis
https://stackoverflow.com/questions/15079327/amortized-complexity-in-laymans-terms

Reply:
ありがとうございます。
いただいた記事を元に調べ、償却計算量という言葉を初めて知りました。
⇒同じ処理をN回行った時に、1回あたりにかかる計算量　と理解しました。

2倍のサイズでコピーした時の計算量が 1 + 2 + 4 + ... 2^(logn) = O(n) となり、
1回あたりの push() の計算量は O(1) になる ということだったのですね。